distributed/layers.py

```diff
@@ -43,20 +61,15 @@ def __init__(
         out_dim_local = out_dim // comm_out_size
 
         # parameters
-        # weights are shared on all comm dims other than the ones used (comm_inp_name, comm_out_name)
-        comm_names_shared = [c for c in comm.get_names(meta=False) if c not in [comm_inp_name, comm_out_name]]
         self.weight = nn.Parameter(torch.ones(out_dim_local, inp_dim_local))
-        self.weight.is_shared_mp = comm_names_shared
-        self.weight.sharded_dims_mp = [
-            self.comm_out_name,
-            self.comm_inp_name,
-            None,
-            None,
-        ]
+        self.weight.is_shared_mp = [comm_act_name] # weights are sharded in tp but shared across cp
+        self.weight.mark_for_reduction = [comm_act_name] # shared weights must be additionally reduced
         if bias:
             self.bias = nn.Parameter(torch.ones(1, 1, out_dim_local))
-            self.bias.is_shared_mp = comm_names_shared
-            self.bias.sharded_dims_mp = [None, self.comm_out_name, None, None]
+            # if inp dim of W is sharded, then the bias is shared across this group and also
+            # shared in cp grp
+            self.bias.is_shared_mp = [self.comm_inp_name, comm_act_name]
```

Comment:
The bias should only be reduced over comm_act_name, so we just can add this to the is_shared_mp variable. We do not need the second one I think, or are we using that in a different context? For initial parameter broadcasting?

Reply:
Yeah for init param syncs (see previous comment)